### PR TITLE
kube-apiserver: Adapted initial delay seconds of livenessProbe to livez-grace-period

### PR DESCRIPTION
**Pull request by mganter**
**How to categorize this PR?**

/area control-plane
/kind cleanup

**What this PR does / why we need it**:
This pull request adapts the initial delay of the kube-apiservers liveness probe to the grace period of the livez endpoint defined in the kube-apiserver args.  

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:
Relates to  https://github.com/gardener/gardener/blob/7145aa25f9523945ec7bfbd257ec9101176ade4c/charts/seed-controlplane/charts/kube-apiserver/templates/deployment.yaml#L142

**Release note**:

```other operator
Liveness probe of kube-apiserver is now adapted to the grace period of the /livez endpoint.
```

### CODE REVIEW

**Comment by gardener-robot-ci-2** [comment]
Thank you @mganter for your contribution. Before I can start building your PR, a member of the organization must set the required label(s) {'reviewed/ok-to-test'}. Once started, you can check the build status in the PR checks section below.

**Comment by rfranzke** [comment]
Thanks @mganter!
/reviewed ok-to-test

Can you please explain the rationale of this change? Why do we need it/what is currently happening without this change?

**Comment by mganter** [comment]
Thanks as well :D

the liveness probe can now run into the live grace period which accepts unfinished post-start hooks before terminating the application.

Therefor the apiserver is considered alive even if there are some post-start hooks have not completed, yet.
Ping, log, etcd and autoregister-completion checks won't be ignored.

From --livez-grace-period duration
`This option represents the maximum amount of time it should take for apiserver to complete its startup sequence and become live. From apiserver's start time to when this amount of time has elapsed, /livez will assume that unfinished post-start hooks will complete successfully and therefore return true.`

Example /livez and the functionality it ensures.
```
[+]ping ok
[+]log ok
[+]etcd ok
[+]poststarthook/start-kube-apiserver-admission-initializer ok
[+]poststarthook/generic-apiserver-start-informers ok
[+]poststarthook/max-in-flight-filter ok
[+]poststarthook/start-apiextensions-informers ok
[+]poststarthook/start-apiextensions-controllers ok
[+]poststarthook/crd-informer-synced ok
[+]poststarthook/bootstrap-controller ok
[+]poststarthook/rbac/bootstrap-roles ok
[+]poststarthook/scheduling/bootstrap-system-priority-classes ok
[+]poststarthook/start-cluster-authentication-info-controller ok
[+]poststarthook/aggregator-reload-proxy-client-cert ok
[+]poststarthook/start-kube-aggregator-informers ok
[+]poststarthook/apiservice-registration-controller ok
[+]poststarthook/apiservice-status-available-controller ok
[+]poststarthook/kube-apiserver-autoregistration ok
[+]autoregister-completion ok
[+]poststarthook/apiservice-openapi-controller ok
healthz check passed
```

**Comment by rfranzke** [comment]
OK, thanks, I see. So, without this change, we might report `kube-apiserver` pods as "healthy" too early (after `15s`) although some of the post-start hooks are not yet completed. Consequently, it's better to wait for the `1m` (livez-grace-period) before starting the ready checks because after this time non-finished post-start hooks are indeed reported as "unhealthy", right?

**Comment by mganter** [comment]
That's exactly what we want to achive

File: charts/seed-controlplane/charts/kube-apiserver/templates/deployment.yaml [hunk 1]
```diff
@@ -194,7 +194,11 @@ spec:
               value: Bearer {{ .Values.probeCredentials }}
           successThreshold: 1
           failureThreshold: 3
+          {{- if semverCompare ">= 1.16" .Values.kubernetesVersion }}
+          initialDelaySeconds: 60
```
Comment (timebertt) [inline charts/seed-controlplane/charts/kube-apiserver/templates/deployment.yaml:198]: I'm confused. How does changing the `initialDelaySeconds` of the liveness probe help in achieving the above goal (delaying the ready status)?
The liveness probe is not considered for the ready status, if I got it correctly. Also see https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/.

E.g., when applying this Pod:
```yaml
apiVersion: v1
kind: Pod
metadata:
  labels:
    test: liveness
  name: liveness-exec
spec:
  terminationGracePeriodSeconds: 1
  containers:
  - name: liveness
    image: k8s.gcr.io/busybox
    args:
    - /bin/sh
    - -c
    - sleep 30; touch /tmp/healthy; sleep 30; rm -rf /tmp/healthy; sleep 600
    livenessProbe:
      exec:
        command:
        - cat
        - /tmp/healthy
      initialDelaySeconds: 30
      periodSeconds: 5
    readinessProbe:
      exec:
        command:
        - echo
        - ready
      initialDelaySeconds: 5
      periodSeconds: 5
```
it gets ready after the 5 seconds delay for the readiness probe even though the liveness probe is delayed for 30s.

```
$ k get po -w
NAME            READY   STATUS    RESTARTS   AGE
liveness-exec   0/1     Pending   0          0s
liveness-exec   0/1     Pending   0          0s
liveness-exec   0/1     ContainerCreating   0          0s
liveness-exec   0/1     ContainerCreating   0          1s
liveness-exec   0/1     Running             0          2s
liveness-exec   1/1     Running             0          8s
```

**Comment by mganter** [comment]
HI there, we did some tests regarding livez-grace-period, it seems like it doesn't have any effect. I will close this PR until i have more accurate information.

**Pull request closed without merging**